File: internal/pricer/pricer_test.go
package pricer

import (
	"math"
	"math/big"
	"testing"
)

func robinhoodState() *State {
	return &State{
		Constraints: []Constraint{
			{Target: 60_000_000, Window: 15, Backlog: 3_111_506},
			{Target: 40_000_000, Window: 86_400, Backlog: 11_194_391_810_886},
		},
		MinBaseFee: big.NewInt(20_000_000),
	}
}

func TestApproxExpBips(t *testing.T) {
	cases := []struct {
		x        Bips
		accuracy uint64
		want     Bips
	}{
		{0, 4, 10_000},
		{10_000, 4, 27_083},
		{32_425, 4, 197_863},
		{0, 0, 10_000},
		{-10_000, 4, 3_692}, // 10_000 * 10_000 / 27_083
	}
	for _, tc := range cases {
		if got := ApproxExpBips(tc.x, tc.accuracy); got != tc.want {
			t.Errorf("ApproxExpBips(%d, %d) = %d, want %d", tc.x, tc.accuracy, got, tc.want)
		}
	}
	// Saturating multiplication keeps huge inputs finite, exactly as nitro does:
	// the last iteration is b + MaxUint64/b.
	if got := ApproxExpBips(Bips(math.MaxInt64), 4); got != Bips(10_000+uint64(math.MaxUint64)/10_000) {
		t.Errorf("saturation: got %d", got)
	}
	// The polynomial is a truncated exp: 1 + x + x^2/2 + x^3/6 + x^4/24.
	for _, x := range []Bips{5_000, 20_000, 32_425} {
		fx := float64(x) / 10_000
		want := 1 + fx + fx*fx/2 + fx*fx*fx/6 + fx*fx*fx*fx/24
		got := float64(ApproxExpBips(x, 4)) / 10_000
		if math.Abs(got-want)/want > 0.001 {
			t.Errorf("ApproxExpBips(%d) = %f, taylor = %f", x, got, want)
		}
	}
}

func TestStepRobinhoodVector(t *testing.T) {
	s := robinhoodState()
	baseFee, exponent, per := s.Step(0)
	if exponent != 32_425 {
		t.Fatalf("exponent = %d, want 32425", exponent)
	}
	if per[0] != 34 || per[1] != 32_391 {
		t.Fatalf("per constraint = %v, want [34 32391]", per)
	}
	if baseFee.Cmp(big.NewInt(395_726_000)) != 0 {
		t.Fatalf("base fee = %s, want 395726000", baseFee)
	}
	observed := big.NewInt(399_726_000)
	if e := ErrorBips(baseFee, observed); e > 200 {
		t.Fatalf("prediction differs from observed by %d bips, want < 200", e)
	}
	// Backlogs unchanged with dt = 0.
	if got := s.Backlogs(); got[0] != 3_111_506 || got[1] != 11_194_391_810_886 {
		t.Fatalf("backlogs = %v", got)
	}
}

func TestStepDrainsAndAddsGas(t *testing.T) {
	s := robinhoodState()
	s.Step(1)
	if got := s.Backlogs(); got[0] != 0 || got[1] != 11_194_391_810_886-40_000_000 {
		t.Fatalf("after dt=1: %v", got)
	}
	s.AddGas(1_000)
	if got := s.Backlogs(); got[0] != 1_000 || got[1] != 11_194_391_810_886-40_000_000+1_000 {
		t.Fatalf("after AddGas: %v", got)
	}
	// Zero backlog and zero divisor contribute nothing and do not panic.
	s2 := &State{Constraints: []Constraint{{Target: 0, Window: 0, Backlog: 5}}, MinBaseFee: big.NewInt(7)}
	fee, exp, per := s2.Step(0)
	if fee.Int64() != 7 || exp != 0 || per[0] != 0 {
		t.Fatalf("zero divisor: fee=%s exp=%d per=%v", fee, exp, per)
	}
	// Nil MinBaseFee is treated as zero.
	s3 := &State{Constraints: []Constraint{{Target: 1, Window: 1, Backlog: 5}}}
	if fee, _, _ := s3.Step(0); fee.Sign() != 0 {
		t.Fatalf("nil min base fee: %s", fee)
	}
}

func TestLegacyModel(t *testing.T) {
	// Robinhood testnet style parameters: speed limit 7M, inertia 102, tolerance 10.
	s := &State{
		Legacy:     &Legacy{SpeedLimit: 7_000_000, Inertia: 102, Tolerance: 10, Backlog: 0},
		MinBaseFee: big.NewInt(10_000_000),
	}
	if !s.IsLegacy() {
		t.Fatal("expected legacy")
	}
	fee, exp, per := s.Step(0)
	if fee.Int64() != 10_000_000 || exp != 0 || per[0] != 0 {
		t.Fatalf("at floor: fee=%s exp=%d per=%v", fee, exp, per)
	}
	// Push the backlog above tolerance*speedLimit (70M) by 714M gas: excess
	// 714M over inertia*speedLimit 714M gives exponent 10_000 bips.
	s.AddGas(70_000_000 + 714_000_000)
	fee, exp, per = s.Step(0)
	if exp != 10_000 || per[0] != 10_000 {
		t.Fatalf("exponent = %d per=%v, want 10000", exp, per)
	}
	if fee.Int64() != 27_083_000 {
		t.Fatalf("fee = %s, want 27083000", fee)
	}
	// Draining for 2 s removes 14M gas.
	s.Step(2)
	if got := s.Backlogs(); got[0] != 784_000_000-14_000_000 {
		t.Fatalf("backlog after drain = %v", got)
	}
	s.SetBacklogs([]uint64{5})
	if s.Backlogs()[0] != 5 {
		t.Fatal("SetBacklogs on legacy failed")
	}
	// Zero inertia never divides by zero.
	z := &State{Legacy: &Legacy{SpeedLimit: 1, Inertia: 0, Tolerance: 0, Backlog: 100}, MinBaseFee: big.NewInt(3)}
	if fee, exp, _ := z.Step(0); fee.Int64() != 3 || exp != 0 {
		t.Fatalf("zero inertia: %s %d", fee, exp)
	}
	// Legacy state without parameters is the floor.
	e := &State{MinBaseFee: big.NewInt(9)}
	if fee, exp, per := e.Step(1); fee.Int64() != 9 || exp != 0 || per != nil {
		t.Fatalf("empty state: %s %d %v", fee, exp, per)
	}
	e.AddGas(1)
	e.SetBacklogs([]uint64{1})
	if e.Backlogs() != nil {
		t.Fatal("empty state has no backlogs")
	}
}

func TestSaturation(t *testing.T) {
	if SaturatingUAdd(math.MaxUint64, 1) != math.MaxUint64 {
		t.Fatal("UAdd")
	}
	if SaturatingUSub(1, 2) != 0 || SaturatingUSub(5, 2) != 3 {
		t.Fatal("USub")
	}
	if SaturatingUMul(math.MaxUint64, 2) != math.MaxUint64 || SaturatingUMul(0, 5) != 0 || SaturatingUMul(3, 4) != 12 {
		t.Fatal("UMul")
	}
	if NaturalToBips(math.MaxUint64) != Bips(math.MaxInt64) {
		t.Fatal("NaturalToBips")
	}
	if SaturatingAddBips(Bips(math.MaxInt64), 1) != Bips(math.MaxInt64) {
		t.Fatal("AddBips overflow")
	}
	if SaturatingAddBips(Bips(math.MinInt64), -1) != Bips(math.MinInt64) {
		t.Fatal("AddBips underflow")
	}
	if SaturatingAddBips(1, 2) != 3 {
		t.Fatal("AddBips")
	}
	// A huge backlog saturates the exponent and the multiplier without panicking.
	s := &State{Constraints: []Constraint{{Target: 1, Window: 1, Backlog: math.MaxUint64}}, MinBaseFee: big.NewInt(1)}
	s.AddGas(math.MaxUint64)
	fee, exp, _ := s.Step(0)
	if exp != Bips(math.MaxInt64) || fee.Sign() <= 0 {
		t.Fatalf("saturated step: exp=%d fee=%s", exp, fee)
	}
	if s.Backlogs()[0] != math.MaxUint64 {
		t.Fatal("AddGas should saturate")
	}
	l := &State{Legacy: &Legacy{SpeedLimit: 1, Inertia: 1, Tolerance: 0, Backlog: math.MaxUint64}, MinBaseFee: big.NewInt(1)}
	l.AddGas(1)
	if l.Backlogs()[0] != math.MaxUint64 {
		t.Fatal("legacy AddGas should saturate")
	}
}

func TestClone(t *testing.T) {
	s := robinhoodState()
	s.Legacy = &Legacy{SpeedLimit: 1}
	c := s.Clone()
	c.Constraints[0].Backlog = 1
	c.Legacy.SpeedLimit = 2
	c.MinBaseFee.SetInt64(1)
	if s.Constraints[0].Backlog == 1 || s.Legacy.SpeedLimit == 2 || s.MinBaseFee.Int64() == 1 {
		t.Fatal("clone shares state")
	}
	empty := (&State{}).Clone()
	if empty.MinBaseFee == nil || empty.Constraints != nil || empty.Legacy != nil {
		t.Fatalf("empty clone: %+v", empty)
	}
}

func TestErrorBips(t *testing.T) {
	if ErrorBips(big.NewInt(101), big.NewInt(100)) != 100 {
		t.Fatal("1% should be 100 bips")
	}
	if ErrorBips(big.NewInt(99), big.NewInt(100)) != 100 {
		t.Fatal("abs")
	}
	if ErrorBips(big.NewInt(1), nil) != 0 || ErrorBips(nil, big.NewInt(1)) != 0 || ErrorBips(big.NewInt(1), big.NewInt(0)) != 0 {
		t.Fatal("nil/zero")
	}
	huge := new(big.Int).Lsh(big.NewInt(1), 80)
	if ErrorBips(huge, big.NewInt(1)) != math.MaxInt64 {
		t.Fatal("saturate")
	}
}

func TestReplay(t *testing.T) {
	s := robinhoodState()
	blocks := []Block{
		{Number: 100, Timestamp: 1000, GasUsed: 10_000_000, BaseFee: big.NewInt(395_726_000)},
		{Number: 101, Timestamp: 1000, GasUsed: 5_000_000, BaseFee: big.NewInt(396_000_000)},
		{Number: 102, Timestamp: 1001, GasUsed: 0, BaseFee: big.NewInt(0)},
	}
	anchor := func(n uint64) ([]uint64, bool) {
		if n == 101 {
			return []uint64{1, 2}, true
		}
		return nil, false
	}
	res := Replay(s, 0, blocks, anchor)
	if len(res) != 3 {
		t.Fatalf("results = %d", len(res))
	}
	// Block 100: dt=0, predicted from the vector, then 10M gas added.
	if res[0].Predicted.Int64() != 395_726_000 || res[0].ErrorBips != 0 || res[0].Exponent != 32_425 {
		t.Fatalf("block 100: %+v", res[0])
	}
	if res[0].Backlogs[0] != 13_111_506 || res[0].Anchored {
		t.Fatalf("block 100 backlogs: %+v", res[0])
	}
	// Block 101: dt=0, exponent grew, anchored afterwards.
	if res[1].Exponent <= 32_425 || !res[1].Anchored || res[1].Backlogs[0] != 1 || res[1].Backlogs[1] != 2 {
		t.Fatalf("block 101: %+v", res[1])
	}
	if res[1].ErrorBips == 0 {
		t.Fatalf("block 101 should record an error against 396000000: %+v", res[1])
	}
	// Block 102: dt=1 drains the tiny anchored backlogs to zero, fee at floor.
	if res[2].Predicted.Int64() != 20_000_000 || res[2].Backlogs[0] != 0 || res[2].Backlogs[1] != 0 || res[2].ErrorBips != 0 {
		t.Fatalf("block 102: %+v", res[2])
	}
	// Continuing from a previous timestamp applies dt to the first block.
	s2 := robinhoodState()
	res2 := Replay(s2, 999, blocks[:1], nil)
	if res2[0].Backlogs[0] != 10_000_000 {
		t.Fatalf("prevTimestamp not applied: %+v", res2[0])
	}
	if Replay(s2, 0, nil, nil) == nil {
		t.Fatal("empty replay should return an empty slice")
	}
}

File: web/src/lib/pricer.test.ts
import { describe, expect, it } from "vitest";
import {
  addGas,
  approxExpBips,
  baseFeeFromExponent,
  constraintExponentBips,
  contributionsBips,
  exponentBips,
  legacyAddGas,
  legacyExponentBips,
  legacyStep,
  MAX_INT64,
  MAX_UINT64,
  MIN_INT64,
  multiplierBips,
  naturalToBips,
  p4Multiplier,
  saturatingAddBips,
  saturatingCastToBips,
  saturatingUAdd,
  saturatingUMul,
  saturatingUSub,
  step,
  stepBacklogs,
  toLegacyState,
  toState,
  toUint64,
  trueExpMultiplier,
} from "./pricer";

describe("approxExpBips (vectors shared with internal/pricer TestApproxExpBips)", () => {
  it("matches nitro at x = 0 and x = 1", () => {
    expect(approxExpBips(0n)).toBe(10_000n);
    // 10000 + 10000/4 = 12500; 10000 + 12500*10000/30000 = 14166;
    // 10000 + 14166*10000/20000 = 17083; 10000 + 17083*10000/10000 = 27083.
    expect(approxExpBips(10_000n)).toBe(27_083n);
    expect(approxExpBips(32_425n)).toBe(197_863n);
  });
  it("is exactly 1 at accuracy 0 instead of dividing by zero", () => {
    expect(approxExpBips(0n, 0)).toBe(10_000n);
    expect(approxExpBips(50_000n, 0)).toBe(10_000n);
  });
  it("takes the reciprocal branch for negative exponents", () => {
    // 10_000 * 10_000 / 27_083.
    expect(approxExpBips(-10_000n)).toBe(3_692n);
    expect(approxExpBips(MIN_INT64)).toBeGreaterThanOrEqual(0n);
  });
  it("saturates huge inputs exactly as Go: the last iteration is b + MaxUint64/b", () => {
    expect(approxExpBips(MAX_INT64)).toBe(10_000n + MAX_UINT64 / 10_000n);
  });
  it("is a truncated exp: 1 + x + x^2/2 + x^3/6 + x^4/24", () => {
    for (const x of [5_000n, 20_000n, 32_425n]) {
      const fx = Number(x) / 10_000;
      const want = 1 + fx + fx ** 2 / 2 + fx ** 3 / 6 + fx ** 4 / 24;
      const got = Number(approxExpBips(x)) / 10_000;
      expect(Math.abs(got - want) / want).toBeLessThan(0.001);
    }
    expect(Number(approxExpBips(50_000n)) / 10_000).toBeLessThan(Math.exp(5));
    expect(approxExpBips(20_000n, 1)).toBe(30_000n);
  });
});

describe("saturating helpers (vectors shared with internal/pricer TestSaturation)", () => {
  it("add, sub and mul saturate at the uint64 bounds", () => {
    expect(saturatingUAdd(MAX_UINT64, 1n)).toBe(MAX_UINT64);
    expect(saturatingUAdd(1n, 2n)).toBe(3n);
    expect(saturatingUSub(1n, 2n)).toBe(0n);
    expect(saturatingUSub(5n, 2n)).toBe(3n);
    expect(saturatingUMul(MAX_UINT64, 2n)).toBe(MAX_UINT64);
    expect(saturatingUMul(0n, 5n)).toBe(0n);
    expect(saturatingUMul(5n, 0n)).toBe(0n);
    expect(saturatingUMul(3n, 4n)).toBe(12n);
  });
  it("casts to bips saturating at MaxInt64", () => {
    expect(naturalToBips(MAX_UINT64)).toBe(MAX_INT64);
    expect(naturalToBips(5n)).toBe(50_000n);
    expect(saturatingCastToBips(MAX_INT64 + 1n)).toBe(MAX_INT64);
    expect(saturatingAddBips(MAX_INT64, 1n)).toBe(MAX_INT64);
    expect(saturatingAddBips(MIN_INT64, -1n)).toBe(MIN_INT64);
    expect(saturatingAddBips(1n, 2n)).toBe(3n);
  });
  it("a valid backlog above MaxUint64 / 10_000 saturates NaturalToBips as Go does", () => {
    // 2e18 * 10_000 overflows uint64, so the numerator is MaxInt64: 2_668_799 bips, not 5_787_037.
    const c = { target: 40_000_000n, window: 86_400n, backlog: 2_000_000_000_000_000_000n };
    expect(constraintExponentBips(c)).toBe(2_668_799n);
  });
  it("a huge backlog saturates the exponent and the multiplier without failing", () => {
    const c = toState([{ target: 1, window: 1, backlog: Number.MAX_SAFE_INTEGER }]);
    c[0].backlog = MAX_UINT64;
    addGas(c, MAX_UINT64);
    expect(c[0].backlog).toBe(MAX_UINT64);
    const r = step(c, 0n, 1n);
    expect(r.exponent).toBe(MAX_INT64);
    expect(r.baseFee).toBeGreaterThan(0n);
    expect(exponentBips([c[0], c[0]])).toBe(MAX_INT64);
    const l = toLegacyState({ speedLimit: 1, inertia: 1, tolerance: 0, backlog: 0 });
    l.backlog = MAX_UINT64;
    legacyAddGas(l, 1n);
    expect(l.backlog).toBe(MAX_UINT64);
  });
  it("converts JSON numbers into uint64 with clamping", () => {
    expect(toUint64(-5)).toBe(0n);
    expect(toUint64(Number.NaN)).toBe(0n);
    expect(toUint64(7.9)).toBe(7n);
    expect(toUint64(2 ** 70)).toBe(MAX_UINT64);
  });
});

describe("Robinhood replay vector (SPEC Appendix B, internal/pricer TestStepRobinhoodVector)", () => {
  const constraints = toState([
    { target: 60_000_000, window: 15, backlog: 3_111_506 },
    { target: 40_000_000, window: 86_400, backlog: 11_194_391_810_886 },
  ]);
  const minBaseFee = 20_000_000n;

  it("sums per-constraint contributions with integer division", () => {
    expect(constraintExponentBips(constraints[0])).toBe(34n);
    expect(constraintExponentBips(constraints[1])).toBe(32_391n);
    expect(exponentBips(constraints)).toBe(32_425n);
    expect(contributionsBips([
      { target: 60_000_000, window: 15, backlog: 3_111_506 },
      { target: 40_000_000, window: 86_400, backlog: 11_194_391_810_886 },
    ])).toEqual([34, 32_391]);
  });

  it("prices within 2% of the observed 399726000 wei", () => {
    const fee = baseFeeFromExponent(minBaseFee, exponentBips(constraints));
    const observed = 399_726_000;
    const error = Math.abs(Number(fee) - observed) / observed;
    expect(error).toBeLessThan(0.02);
    expect(fee).toBe(395_726_000n);
    expect(multiplierBips(32_425n)).toBe(197_863n);
  });

  it("returns the floor when the exponent is zero", () => {
    expect(baseFeeFromExponent(minBaseFee, 0n)).toBe(minBaseFee);
    expect(multiplierBips(0n)).toBe(10_000n);
    expect(constraintExponentBips({ target: 1n, window: 0n, backlog: 5n })).toBe(0n);
    expect(constraintExponentBips({ target: 0n, window: 0n, backlog: 5n })).toBe(0n);
  });

  it("step with dt = 0 leaves the backlogs alone and reports the vector", () => {
    const c = toState([
      { target: 60_000_000, window: 15, backlog: 3_111_506 },
      { target: 40_000_000, window: 86_400, backlog: 11_194_391_810_886 },
    ]);
    const r = step(c, 0n, minBaseFee);
    expect(r.exponent).toBe(32_425n);
    expect(r.contributions).toEqual([34n, 32_391n]);
    expect(r.baseFee).toBe(395_726_000n);
    expect(c.map((x) => x.backlog)).toEqual([3_111_506n, 11_194_391_810_886n]);
  });
});

describe("step and addGas", () => {
  it("drains at target rate, saturating at zero, then prices", () => {
    const c = toState([
      { target: 60_000_000, window: 15, backlog: 100_000_000 },
      { target: 40_000_000, window: 86_400, backlog: 3_456_000_000_000 },
    ]);
    const r = step(c, 1n, 20_000_000n);
    expect(c[0].backlog).toBe(40_000_000n);
    expect(c[1].backlog).toBe(3_456_000_000_000n - 40_000_000n);
    expect(r.contributions[0]).toBe(444n);
    expect(r.exponent).toBe(444n + r.contributions[1]);
    expect(r.baseFee).toBeGreaterThan(20_000_000n);
    const again = step(c, 10n, 20_000_000n);
    expect(c[0].backlog).toBe(0n);
    expect(again.contributions[0]).toBe(0n);
    addGas(c, 5n);
    expect(c[0].backlog).toBe(5n);
  });

  it("matches TestStepDrainsAndAddsGas", () => {
    const c = toState([
      { target: 60_000_000, window: 15, backlog: 3_111_506 },
      { target: 40_000_000, window: 86_400, backlog: 11_194_391_810_886 },
    ]);
    step(c, 1n, 20_000_000n);
    expect(c.map((x) => x.backlog)).toEqual([0n, 11_194_391_810_886n - 40_000_000n]);
    addGas(c, 1_000n);
    expect(c.map((x) => x.backlog)).toEqual([1_000n, 11_194_391_810_886n - 40_000_000n + 1_000n]);
    const zero = step([{ target: 0n, window: 0n, backlog: 5n }], 0n, 7n);
    expect(zero.baseFee).toBe(7n);
    expect(zero.exponent).toBe(0n);
    expect(zero.contributions).toEqual([0n]);
  });

  it("stepBacklogs drains with fractional dt and never goes negative", () => {
    const constraints = [
      { target: 60_000_000, window: 15, backlog: 30_000_000, exponentBips: 0 },
      { target: 40_000_000, window: 86_400, backlog: 1_000_000_000, exponentBips: 0 },
    ];
    expect(stepBacklogs(constraints, 0.25)).toEqual([15_000_000, 990_000_000]);
    expect(stepBacklogs(constraints, 10)).toEqual([0, 600_000_000]);
    expect(stepBacklogs(constraints, -1)).toEqual([30_000_000, 1_000_000_000]);
  });

  it("toState clamps negative backlogs", () => {
    expect(toState([{ target: 1, window: 1, backlog: -5 }])[0].backlog).toBe(0n);
  });
});

describe("comparison helpers", () => {
  it("evaluates P4 and e^x", () => {
    expect(p4Multiplier(0)).toBe(1);
    expect(p4Multiplier(1)).toBeCloseTo(2.7083, 4);
    expect(p4Multiplier(3.2426)).toBeCloseTo(19.79, 1);
    expect(trueExpMultiplier(3.2426)).toBeCloseTo(25.6, 1);
  });
});

describe("legacy model (internal/pricer TestLegacyModel)", () => {
  it("is at the floor until backlog exceeds tolerance * speedLimit", () => {
    const s = toLegacyState({ speedLimit: 7_000_000, inertia: 102, tolerance: 10, backlog: 60_000_000 });
    expect(legacyExponentBips(s)).toBe(0n);
    const r = legacyStep(s, 0n, 10_000_000n);
    expect(r.baseFee).toBe(10_000_000n);
    legacyAddGas(s, 100_000_000n);
    expect(s.backlog).toBe(160_000_000n);
    // (160M - 70M) * 10000 / (102 * 7M) = 1260 bips
    expect(legacyExponentBips(s)).toBe(1260n);
    const priced = legacyStep(s, 1n, 10_000_000n);
    expect(s.backlog).toBe(153_000_000n);
    expect(priced.exponent).toBe(1162n);
    expect(priced.baseFee).toBeGreaterThan(10_000_000n);
    legacyStep(s, 1000n, 10_000_000n);
    expect(s.backlog).toBe(0n);
    expect(legacyExponentBips({ speedLimit: 0n, inertia: 1n, tolerance: 0n, backlog: 5n })).toBe(0n);
    expect(toLegacyState({ speedLimit: 1, inertia: 1, tolerance: 1, backlog: -2 }).backlog).toBe(0n);
  });
  it("matches the Go vector: 784M gas over a 70M threshold with 714M inertia is exactly 10_000 bips", () => {
    const s = toLegacyState({ speedLimit: 7_000_000, inertia: 102, tolerance: 10, backlog: 0 });
    legacyAddGas(s, 70_000_000n + 714_000_000n);
    const r = legacyStep(s, 0n, 10_000_000n);
    expect(r.exponent).toBe(10_000n);
    expect(r.baseFee).toBe(27_083_000n);
    legacyStep(s, 2n, 10_000_000n);
    expect(s.backlog).toBe(784_000_000n - 14_000_000n);
    // Zero inertia never divides by zero.
    const z = toLegacyState({ speedLimit: 1, inertia: 0, tolerance: 0, backlog: 100 });
    expect(legacyStep(z, 0n, 3n)).toEqual({ baseFee: 3n, exponent: 0n });
  });
});
